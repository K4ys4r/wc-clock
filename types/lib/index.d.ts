export { ClockComponent };
import { ClockComponent } from "./wc-clock.js";
//# sourceMappingURL=index.d.ts.map