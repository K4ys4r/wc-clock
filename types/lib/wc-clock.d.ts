export class ClockComponent extends HTMLElement {
    static get tagName(): string;
    static get attributes(): {
        "minutes-ticks": string;
        "hours-ticks": string;
        "time-zone": string;
        "fixed-time": string;
        "hours-labels": string;
    };
    static get observedAttributes(): any;
    hoursLabels: any;
    connectedCallback(): void;
    disconnectedCallback(): void;
    attributeChangedCallback(name: any, oldValue: any, newValue: any): void;
    timeZone: any;
    fixedTime: any;
    #private;
}
//# sourceMappingURL=wc-clock.d.ts.map