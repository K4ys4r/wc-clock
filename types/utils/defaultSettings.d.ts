export namespace defaultSettings {
    const componentName: string;
    const componentAttributes: {
        "minutes-ticks": string;
        "hours-ticks": string;
        "time-zone": string;
        "fixed-time": string;
        "hours-labels": string;
    };
    const componentTemplate: HTMLTemplateElement;
    const hoursLabels: any;
    const fixedTime: number[];
    const timeZone: string;
}
//# sourceMappingURL=defaultSettings.d.ts.map