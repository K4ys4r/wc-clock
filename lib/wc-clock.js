import { defaultSettings } from "../utils/index.js";

export class ClockComponent extends HTMLElement {
  #interval;
  constructor() {
    super();
    this.attachShadow({ mode: "open" });
    const temp = defaultSettings.componentTemplate.content.cloneNode(true);
    this.shadowRoot.append(temp);
    this.hoursLabels = defaultSettings.hoursLabels;
  }
  static get tagName() {
    return defaultSettings.componentName;
  }
  static get attributes() {
    return defaultSettings.componentAttributes;
  }
  static get observedAttributes() {
    return Object.values(ClockComponent.attributes);
  }
  connectedCallback() {
    this.#render();
  }

  disconnectedCallback() {
    clearInterval(this.#interval);
  }

  attributeChangedCallback(name, oldValue, newValue) {
    switch (name) {
      case ClockComponent.attributes["hours-labels"]:
        this.hoursLabels =
          newValue === ""
            ? defaultSettings.hoursLabels
            : newValue
            ? newValue.split(",")
            : [];
        this.#setHoursLabels();
        break;
      case ClockComponent.attributes["minutes-ticks"]:
        this.#setTicks(
          newValue === ""
            ? Array.from({ length: 60 }, (_, i) => {
                if (i % 5 != 0) return i;
              }).filter(Number)
            : [],
          false
        );
        break;
      case ClockComponent.attributes["hours-ticks"]:
        this.#setTicks(
          newValue === "" ? Array.from({ length: 12 }, (_, i) => i * 5) : []
        );
        break;
      case ClockComponent.attributes["time-zone"]:
        this.timeZone = Intl.supportedValuesOf("timeZone").includes(newValue)
          ? newValue
          : defaultSettings.timeZone;
        this.#render();
        break;
      case ClockComponent.attributes["fixed-time"]:
        this.fixedTime =
          newValue === ""
            ? defaultSettings.fixedTime
            : newValue
            ? newValue.split(":").map(Number) || null
            : null;
        this.#render();
        break;
    }
  }

  #render() {
    clearInterval(this.#interval);
    const hourHand = this.shadowRoot.getElementById(
      "wc-clock__clock-hours-hand"
    );
    const minuteHand = this.shadowRoot.getElementById(
      "wc-clock__clock-minutes-hand"
    );
    const secondHand = this.shadowRoot.getElementById(
      "wc-clock__clock-seconds-hand"
    );
    if (!this.fixedTime) {
      let date = new Date(
        new Date().toLocaleString("en", { timeZone: this.timeZone })
      );

      this.#updateClock({ date, hourHand, minuteHand, secondHand });
      this.#interval = setInterval(() => {
        date = new Date(
          new Date().toLocaleString("en", { timeZone: this.timeZone })
        );
        this.#updateClock({ date, hourHand, minuteHand, secondHand });
      }, 1000);
    } else {
      let date = new Date();
      date.setHours(...this.fixedTime);
      this.#updateClock({ date, hourHand, minuteHand, secondHand });
    }
  }

  #updateClock({ date, hourHand, minuteHand, secondHand }) {
    let hh = date.getHours();
    let mm = date.getMinutes();
    let ss = date.getSeconds();
    let minuteRotate = 6 * mm; // 6deg/min
    let secondRotate = 6 * ss; // 6deg/min
    let hourRotate = 30 * hh + mm / 2; // 30deg/h + 30deg/60min
    hourHand.style.transform = `rotate(${hourRotate}deg)`;
    minuteHand.style.transform = `rotate(${minuteRotate}deg)`;
    secondHand.style.transform = `rotate(${secondRotate}deg)`;
  }

  #setHoursLabels() {
    const symbolsDiv = this.shadowRoot.getElementById("wc-clock__clock-labels");

    // symbolsIndex determine the position of label on the clock based on the length of hoursLabels
    let symbolsIndex = (12 / this.hoursLabels.length) ^ 0;

    // update symbolsDiv elements
    symbolsDiv.replaceChildren(
      ...this.hoursLabels.map((s, i) => {
        let span = document.createElement("span");
        span.innerHTML = `<b>${s}</b>`;
        span.style.setProperty("--a", (i + 1) * symbolsIndex);
        return span;
      })
    );
  }

  #setTicks(numbers, isHours = true) {
    const ticksDiv = this.shadowRoot.getElementById("wc-clock__clock-ticks");

    // get the existing ticks to not delete them
    let oldNodes = ticksDiv.querySelectorAll(
      `[id^=${isHours ? "minutes" : "hours"}-ticks]`
    );

    // update ticksDiv elements
    ticksDiv.replaceChildren(
      ...numbers.map((i) => {
        let span = document.createElement("span");
        span.id = `${isHours ? "hours" : "minutes"}-ticks-${i}`;
        span.style.setProperty("--ta", i);
        span.style.setProperty(
          "--clock-tick-color",
          `var( --clock-${isHours ? "hours" : "minutes"}-ticks-color)`
        );
        span.style.setProperty(
          "--clock-tick-width",
          `var(--clock-${isHours ? "hours" : "minutes"}-ticks-width)`
        );
        span.style.setProperty(
          "--clock-tick-height",
          `var(--clock-${isHours ? "hours" : "minutes"}-ticks-height)`
        );
        return span;
      }),
      ...oldNodes
    );
  }
}
