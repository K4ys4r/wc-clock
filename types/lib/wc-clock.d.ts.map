{"version":3,"file":"wc-clock.d.ts","sourceRoot":"","sources":["../../lib/wc-clock.js"],"names":[],"mappings":"AAEA;IASE,6BAEC;IACD;;;;;;MAEC;IACD,qCAEC;IAVC,iBAA8C;IAWhD,0BAEC;IAED,6BAEC;IAED,wEA0CC;IAfK,cAE4B;IAI5B,eAKU;;CAgGjB"}