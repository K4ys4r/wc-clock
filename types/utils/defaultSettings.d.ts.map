{"version":3,"file":"defaultSettings.d.ts","sourceRoot":"","sources":["../../utils/defaultSettings.js"],"names":[],"mappings":""}