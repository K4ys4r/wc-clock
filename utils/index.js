import { defaultSettings } from "./defaultSettings.js";

export { defaultSettings };
