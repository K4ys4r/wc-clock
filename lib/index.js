import { ClockComponent } from "./wc-clock.js";

customElements.define(ClockComponent.tagName, ClockComponent);
export { ClockComponent };
