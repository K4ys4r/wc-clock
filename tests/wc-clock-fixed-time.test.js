import "../lib/index";
import { ClockComponent } from "../lib/index";
import { beforeEach, describe, expect, it } from "vitest";
import { createWebComponent } from "./utils";

describe("wc-clock", () => {
  beforeEach(async () => {
    document.body.innerHTML = "";
  });
  it("add fixed Time default attribute ", async () => {
    createWebComponent({
      name: ClockComponent.tagName,
      attributes: { "fixed-time": "" },
    });
    const wc = document.querySelector(ClockComponent.tagName);
    expect(wc).not.toBeNull();

    expect(
      wc.getAttribute(ClockComponent.attributes["fixed-time"])
    ).not.toBeNull();
    expect(
      wc.getAttribute(ClockComponent.attributes["hours-labels"])
    ).toBeNull();
    expect(
      wc.getAttribute(ClockComponent.attributes["hours-ticks"])
    ).toBeNull();
    expect(
      wc.getAttribute(ClockComponent.attributes["minutes-ticks"])
    ).toBeNull();

    // test if default fixed time is 10:10:30
    expect(wc.fixedTime).toEqual([10, 10, 30]);
    [
      ["seconds", 180], // 30*6deg
      ["minutes", 60], // 10*6deg
      ["hours", 305], // 10*30deg + 10/2 deg
    ].map(([type, deg]) => {
      let hand = wc.shadowRoot.querySelector(`#wc-clock__clock-${type}-hand`);
      expect(hand).not.toBeNull();
      expect(hand.style.transform).toBe(`rotate(${deg}deg)`);
    });

    ["ticks", "labels"].map((type) => {
      let divType = wc.shadowRoot.querySelector(`#wc-clock__clock-${type}`);
      expect(divType).not.toBeNull();
      expect(divType.children.length).toEqual(0);
    });

    expect(wc.shadowRoot.innerHTML).toMatchSnapshot();
  });
  it("add a specific fixed Time   ", async () => {
    createWebComponent({
      name: ClockComponent.tagName,
      attributes: { "fixed-time": "5:25:55", "hours-ticks": "" },
    });
    // get the web component from the DOM
    const wc = document.querySelector(ClockComponent.tagName);
    expect(wc).not.toBeNull();

    // check its attributes
    expect(
      wc.getAttribute(ClockComponent.attributes["fixed-time"])
    ).not.toBeNull();
    expect(
      wc.getAttribute(ClockComponent.attributes["hours-labels"])
    ).toBeNull();
    expect(
      wc.getAttribute(ClockComponent.attributes["hours-ticks"])
    ).not.toBeNull();
    expect(
      wc.getAttribute(ClockComponent.attributes["minutes-ticks"])
    ).toBeNull();

    // test if  fixed time is 5:25:55
    expect(wc.fixedTime).toEqual([5, 25, 55]);
    [
      ["seconds", 330], // 55*6deg
      ["minutes", 150], // 10*6deg
      ["hours", 162.5], // 10*30deg + 10/2 deg
    ].map(([type, deg]) => {
      let hand = wc.shadowRoot.querySelector(`#wc-clock__clock-${type}-hand`);
      expect(hand).not.toBeNull();
      expect(hand.style.transform).toBe(`rotate(${deg}deg)`);
    });

    // test ticks & labels divs
    [
      ["ticks", 12],
      ["labels", 0],
    ].map(([type, nChild]) => {
      let divType = wc.shadowRoot.querySelector(`#wc-clock__clock-${type}`);
      expect(divType).not.toBeNull();
      expect(divType.children.length).toEqual(nChild);
    });

    // test clock minutes & hours ticks
    [
      ["hours", 12],
      ["minutes", 0],
    ].map(([type, n]) => {
      expect(wc.shadowRoot.querySelectorAll(`[id^=${type}-ticks]`).length).toBe(
        n
      );
    });

    // extract a snapshot
    expect(wc.shadowRoot.innerHTML).toMatchSnapshot();
  });
});
