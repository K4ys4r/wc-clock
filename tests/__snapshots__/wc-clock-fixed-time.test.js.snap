// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wc-clock > add a specific fixed Time    1`] = `
"<style>
  :host {
    display: block;
    --hours-hand-color: #b9b9b9;
    --minutes-hand-color: #b9b9b9;
    --seconds-hand-color: #ff3434;
    --clock-background-color: #1a1a1aff;
    --clock-color: #fff;
    --clock-width: 10rem;
    --seconds-hand-width: 1;
    --seconds-hand-height: 4;
    --minutes-hand-width: 6;
    --minutes-hand-height: 4;
    --hours-hand-width: 6;
    --hours-hand-height: 3;
    --clock-font-size: 1rem;
    --clock-font: Arial, Helvetica, sans-serif;
    --clock-hours-ticks-color: #fff;
    --clock-minutes-ticks-color: #ffffff90;
    --clock-hours-ticks-height: 0.45rem;
    --clock-hours-ticks-width: 0.15rem;
    --clock-minutes-ticks-height: 0.23rem;
    --clock-minutes-ticks-width: 0.05rem;
  }
  .wc-clock__container {
    position: relative;
    display: flex;
    justify-content: center;
    align-items: center;
    font-family: var(--clock-font-family);
    font-size: var(--clock-font-size);
  }
  .wc-clock__clock {
    width: var(--clock-width);
    height: var(--clock-width);
    background: var(--clock-background-color);
    border-radius: 50%;
    box-shadow: 0 0 0.25rem var(--clock-background-color),
      0 0 0.1em 0.2em #a9a9a9;
    border: 0.1em solid var(--clock-background-color);
    display: flex;
    justify-content: center;
    align-items: center;
    .wc-clock__clock-labels {
      > span {
        position: absolute;
        color: var(--clock-color);
        transform: rotate(calc(30deg * var(--a)));
        inset: 0.5em;
        text-align: center;
        > b {
          display: flex;
          justify-content: center;
          transform: rotate(calc(-30deg * var(--a)));
        }
      }
    }
    .wc-clock__clock-ticks {
      > span {
        position: absolute;
        transform: rotate(calc(6deg * var(--ta)));
        text-align: center;
        inset: 0;
        display: flex;
        justify-content: center;
      }
      > span::before {
        content: "";
        width: var(--clock-tick-width);
        height: var(--clock-tick-height);
        border-radius: 0.1em;
        background-color: var(--clock-tick-color, var(--clock-color));
        align-self: flex-start;
      }
    }
  }
  .wc-clock__clock::before {
    position: absolute;
    content: "";
    background-color: #fff;
    width: 1.25rem;
    height: 1.25rem;
    border-radius: 50%;
  }
  .wc-clock__clock::after {
    position: absolute;
    content: "";
    background-color: #fff;
    box-shadow: 0 0 0 0.2rem var(--clock-background-color);
    width: 0.2rem;
    height: 0.2rem;
    border-radius: 50%;
    z-index: 2;
  }

  .wc-clock__clock-hand {
    position: absolute;
    display: flex;
    justify-content: center;
    align-items: flex-end;
    width: 1rem;
    height: 1rem;
    transform: rotate(360deg);
    z-index: 1;
    > i {
      position: absolute;
      width: calc(0.05em * var(--width));
      height: calc(1em * var(--height));
      background: var(--color);
      border-radius: 100% 100% 50% 50%;
      display: flex;
      justify-content: center;
    }
    > i::before {
      content: "";
      position: absolute;
      width: calc(0.05em * var(--width) + 0.1em);
      height: calc(0.05em * var(--height) + 0.5em);
      background: var(--color);
      border-radius: 100% 100% 50% 50%;
      transform: translate(0, 0.25em);
    }
    > i::after {
      content: "";
      position: absolute;
      width: calc(0.05em * var(--width));
      height: calc(0.1em * var(--height));
      background: var(--color);
      border-radius: 1rem 1rem 0.5rem 0.5rem;
      top: 100%;
    }
  }
</style>
<div id="wc-clock__container" class="wc-clock__container">
  <div id="wc-clock__clock" class="wc-clock__clock">
    <div id="wc-clock__clock-minutes-hand" class="wc-clock__clock-hand" style="transform: rotate(150deg);">
      <i style="
          --color: var(--minutes-hand-color);
          --width: var(--minutes-hand-width);
          --height: var(--minutes-hand-height);
        "></i>
    </div>
    <div id="wc-clock__clock-hours-hand" class="wc-clock__clock-hand" style="transform: rotate(162.5deg);">
      <i style="
          --color: var(--hours-hand-color);
          --width: var(--hours-hand-width);
          --height: var(--hours-hand-height);
        "></i>
    </div>
    <div id="wc-clock__clock-seconds-hand" class="wc-clock__clock-hand" style="transform: rotate(330deg);">
      <i style="
          --color: var(--seconds-hand-color);
          --width: var(--seconds-hand-width);
          --height: var(--seconds-hand-height);
        "></i>
    </div>

    <div class="wc-clock__clock-labels" id="wc-clock__clock-labels"></div>
    <div class="wc-clock__clock-ticks" id="wc-clock__clock-ticks"><span id="hours-ticks-0" style="--ta: 0; --clock-tick-color: var( --clock-hours-ticks-color); --clock-tick-width: var(--clock-hours-ticks-width); --clock-tick-height: var(--clock-hours-ticks-height);"></span><span id="hours-ticks-5" style="--ta: 5; --clock-tick-color: var( --clock-hours-ticks-color); --clock-tick-width: var(--clock-hours-ticks-width); --clock-tick-height: var(--clock-hours-ticks-height);"></span><span id="hours-ticks-10" style="--ta: 10; --clock-tick-color: var( --clock-hours-ticks-color); --clock-tick-width: var(--clock-hours-ticks-width); --clock-tick-height: var(--clock-hours-ticks-height);"></span><span id="hours-ticks-15" style="--ta: 15; --clock-tick-color: var( --clock-hours-ticks-color); --clock-tick-width: var(--clock-hours-ticks-width); --clock-tick-height: var(--clock-hours-ticks-height);"></span><span id="hours-ticks-20" style="--ta: 20; --clock-tick-color: var( --clock-hours-ticks-color); --clock-tick-width: var(--clock-hours-ticks-width); --clock-tick-height: var(--clock-hours-ticks-height);"></span><span id="hours-ticks-25" style="--ta: 25; --clock-tick-color: var( --clock-hours-ticks-color); --clock-tick-width: var(--clock-hours-ticks-width); --clock-tick-height: var(--clock-hours-ticks-height);"></span><span id="hours-ticks-30" style="--ta: 30; --clock-tick-color: var( --clock-hours-ticks-color); --clock-tick-width: var(--clock-hours-ticks-width); --clock-tick-height: var(--clock-hours-ticks-height);"></span><span id="hours-ticks-35" style="--ta: 35; --clock-tick-color: var( --clock-hours-ticks-color); --clock-tick-width: var(--clock-hours-ticks-width); --clock-tick-height: var(--clock-hours-ticks-height);"></span><span id="hours-ticks-40" style="--ta: 40; --clock-tick-color: var( --clock-hours-ticks-color); --clock-tick-width: var(--clock-hours-ticks-width); --clock-tick-height: var(--clock-hours-ticks-height);"></span><span id="hours-ticks-45" style="--ta: 45; --clock-tick-color: var( --clock-hours-ticks-color); --clock-tick-width: var(--clock-hours-ticks-width); --clock-tick-height: var(--clock-hours-ticks-height);"></span><span id="hours-ticks-50" style="--ta: 50; --clock-tick-color: var( --clock-hours-ticks-color); --clock-tick-width: var(--clock-hours-ticks-width); --clock-tick-height: var(--clock-hours-ticks-height);"></span><span id="hours-ticks-55" style="--ta: 55; --clock-tick-color: var( --clock-hours-ticks-color); --clock-tick-width: var(--clock-hours-ticks-width); --clock-tick-height: var(--clock-hours-ticks-height);"></span></div>
  </div>
</div>
"
`;

exports[`wc-clock > add fixed Time default attribute  1`] = `
"<style>
  :host {
    display: block;
    --hours-hand-color: #b9b9b9;
    --minutes-hand-color: #b9b9b9;
    --seconds-hand-color: #ff3434;
    --clock-background-color: #1a1a1aff;
    --clock-color: #fff;
    --clock-width: 10rem;
    --seconds-hand-width: 1;
    --seconds-hand-height: 4;
    --minutes-hand-width: 6;
    --minutes-hand-height: 4;
    --hours-hand-width: 6;
    --hours-hand-height: 3;
    --clock-font-size: 1rem;
    --clock-font: Arial, Helvetica, sans-serif;
    --clock-hours-ticks-color: #fff;
    --clock-minutes-ticks-color: #ffffff90;
    --clock-hours-ticks-height: 0.45rem;
    --clock-hours-ticks-width: 0.15rem;
    --clock-minutes-ticks-height: 0.23rem;
    --clock-minutes-ticks-width: 0.05rem;
  }
  .wc-clock__container {
    position: relative;
    display: flex;
    justify-content: center;
    align-items: center;
    font-family: var(--clock-font-family);
    font-size: var(--clock-font-size);
  }
  .wc-clock__clock {
    width: var(--clock-width);
    height: var(--clock-width);
    background: var(--clock-background-color);
    border-radius: 50%;
    box-shadow: 0 0 0.25rem var(--clock-background-color),
      0 0 0.1em 0.2em #a9a9a9;
    border: 0.1em solid var(--clock-background-color);
    display: flex;
    justify-content: center;
    align-items: center;
    .wc-clock__clock-labels {
      > span {
        position: absolute;
        color: var(--clock-color);
        transform: rotate(calc(30deg * var(--a)));
        inset: 0.5em;
        text-align: center;
        > b {
          display: flex;
          justify-content: center;
          transform: rotate(calc(-30deg * var(--a)));
        }
      }
    }
    .wc-clock__clock-ticks {
      > span {
        position: absolute;
        transform: rotate(calc(6deg * var(--ta)));
        text-align: center;
        inset: 0;
        display: flex;
        justify-content: center;
      }
      > span::before {
        content: "";
        width: var(--clock-tick-width);
        height: var(--clock-tick-height);
        border-radius: 0.1em;
        background-color: var(--clock-tick-color, var(--clock-color));
        align-self: flex-start;
      }
    }
  }
  .wc-clock__clock::before {
    position: absolute;
    content: "";
    background-color: #fff;
    width: 1.25rem;
    height: 1.25rem;
    border-radius: 50%;
  }
  .wc-clock__clock::after {
    position: absolute;
    content: "";
    background-color: #fff;
    box-shadow: 0 0 0 0.2rem var(--clock-background-color);
    width: 0.2rem;
    height: 0.2rem;
    border-radius: 50%;
    z-index: 2;
  }

  .wc-clock__clock-hand {
    position: absolute;
    display: flex;
    justify-content: center;
    align-items: flex-end;
    width: 1rem;
    height: 1rem;
    transform: rotate(360deg);
    z-index: 1;
    > i {
      position: absolute;
      width: calc(0.05em * var(--width));
      height: calc(1em * var(--height));
      background: var(--color);
      border-radius: 100% 100% 50% 50%;
      display: flex;
      justify-content: center;
    }
    > i::before {
      content: "";
      position: absolute;
      width: calc(0.05em * var(--width) + 0.1em);
      height: calc(0.05em * var(--height) + 0.5em);
      background: var(--color);
      border-radius: 100% 100% 50% 50%;
      transform: translate(0, 0.25em);
    }
    > i::after {
      content: "";
      position: absolute;
      width: calc(0.05em * var(--width));
      height: calc(0.1em * var(--height));
      background: var(--color);
      border-radius: 1rem 1rem 0.5rem 0.5rem;
      top: 100%;
    }
  }
</style>
<div id="wc-clock__container" class="wc-clock__container">
  <div id="wc-clock__clock" class="wc-clock__clock">
    <div id="wc-clock__clock-minutes-hand" class="wc-clock__clock-hand" style="transform: rotate(60deg);">
      <i style="
          --color: var(--minutes-hand-color);
          --width: var(--minutes-hand-width);
          --height: var(--minutes-hand-height);
        "></i>
    </div>
    <div id="wc-clock__clock-hours-hand" class="wc-clock__clock-hand" style="transform: rotate(305deg);">
      <i style="
          --color: var(--hours-hand-color);
          --width: var(--hours-hand-width);
          --height: var(--hours-hand-height);
        "></i>
    </div>
    <div id="wc-clock__clock-seconds-hand" class="wc-clock__clock-hand" style="transform: rotate(180deg);">
      <i style="
          --color: var(--seconds-hand-color);
          --width: var(--seconds-hand-width);
          --height: var(--seconds-hand-height);
        "></i>
    </div>

    <div class="wc-clock__clock-labels" id="wc-clock__clock-labels"></div>
    <div class="wc-clock__clock-ticks" id="wc-clock__clock-ticks"></div>
  </div>
</div>
"
`;
